src/coreclr/jit/valuenum.cpp

```diff
@@ -7146,9 +7158,9 @@ ValueNum Compiler::fgMemoryVNForLoopSideEffects(MemoryKind  memoryKind,
                 }
 #endif // DEBUG
 
-                // Instance field maps get the placeholder TYP_REF - they do not represent "singular"
+                // Instance field maps get a placeholder type - they do not represent "singular"
                 // values. Static field maps, on the other hand, do, and so must be given proper types.
-                var_types fldMapType = eeIsFieldStatic(fldHnd) ? eeGetFieldType(fldHnd) : TYP_REF;
+                var_types fldMapType = eeIsFieldStatic(fldHnd) ? eeGetFieldType(fldHnd) : TYP_MEM;
```

Comment:
Strictly speaking the part about statics is accurate only for primitive types, right?

Reply:
It is... complicated.

Logically, it is true for all non-shared (shared in the `__Canon` sense) fields, since they have unique identities (with one exception: overlapping RVA statics).

Implementation-wise, it is all over the place, I have a comment about it in the unfinished changes that captures the spirit:
```
// Ordinary statics select their types. "Shared" statics, however, select
// their version of "the first field map", which is then indexed into by
// the logical "instantiation" argument, though see below). Or, at least,
// that was the original intent behind the relevant logic.

// In practice, "IsFieldAddr" "gets things wrong", and we form selection chains
// that are too conservative/long. First, the IR representation for the "boxed
// static" case, on 64 bit, when not using CLS_VAR, makes us think that such fields
// are in fact "shared", and use the "static offset" selector for them. Second, it
// also makes us think the box itself is in fact the field. Third, only fields that
// logically are "shared" (between generic instatiations) need this "static offset",
// but all cases that use helpers will hit it, because their IR shapes are similar or
// identical. For example, for an access of a simple struct static field in R2R code,
// we will form this:
//
//   $staticOffset = CAST(heap[field][helper], ref <- struct)
//   $fieldValue   = heap[field][$staticOffset]
//
// This is not a correctness issue, as we form selections that are "just" too
// conservative, but it does make for a somewhat confusing experience, and leads
// to quite a bit if wasted memory and cycles.
```
You are right that this handling is incomplete, fixing it is WIP.